### PR TITLE
Add GitHub make PR tool function

### PR DESCRIPTION
**Pull request by kgang**
* Add `GithubTool.make_pr`
* Add RESTful HTTP client `RestfulApiClient`
* Rename `DeployToAWS` to `AwsTool`
* Rm `pygithub`
* update `mypy` for more consistency/reliability maybe

### CODE REVIEW

**Comment by MichaelDeng03** [comment]
nice!

File: concrete/tools.py [hunk 1]
```diff
@@ -389,35 +417,24 @@ def _poll_service_status(cls, service_name: str) -> bool:
...
+        }
+
+    def make_pr(self, owner: str, repo: str, branch: str) -> dict:
+        """
+        Make a pull request on the target repo
+
+        e.g. make_pr('abstractoperators', 'concrete', 'kent/http-tool')
+        """
+        url = f"https://api.github.com/repos/{owner}/{repo}/pulls"
+        json = {'title': '[ABOP] Test', 'head': branch, 'base': 'main'}
```
Comment (MichaelDeng03) [inline concrete/tools.py:439]: Rename PR title

File: concrete/tools.py [hunk 2]
```diff
@@ -123,7 +118,39 @@ def __repr__(cls):
...
+
+        Throws an error if the request was unsuccessful after retries
+        """
+        client = RestApiClient()
+        resp = client.get(url, headers=headers, params=params, data=data)
+        if not resp.ok:
+            # Request failed
+            CLIClient.emit(f"Failed GET request to {url}: {resp.status_code} {resp.json()}")
+            resp.raise_for_status()
+        return resp.json()  # return unwrapped data
```
Comment (MichaelDeng03) [inline concrete/tools.py:135]: Should not assume JSON return type

Comment (kgang) [reply concrete/tools.py:135]: For RESTful APIs I think we can

File: concrete/tools.py [hunk 3]
```diff
@@ -123,7 +118,39 @@ def __repr__(cls):
...
+
+        Throws an error if the request was unsuccessful after retries
+        """
+        client = RestApiClient()
+        resp = client.post(url, headers=headers, params=params, data=data, json=json)
+        if not resp.ok:
+            # Request failed
+            CLIClient.emit(f"Failed POST request to {url}: {resp.status_code} {resp.json()}")
+            resp.raise_for_status()
+        return resp.json()  # return unwrapped data
```
Comment (MichaelDeng03) [inline concrete/tools.py:150]: Should not assume json return

File: concrete/tools.py [hunk 4]
```diff
@@ -123,7 +118,39 @@ def __repr__(cls):
...
+
+    @classmethod
+    def post(cls, url: str, headers: dict = {}, params: dict = {}, data: dict = {}, json: dict = {}):
+        """
+        Make a POST request to the specified url
+
+        Throws an error if the request was unsuccessful after retries
+        """
+        client = RestApiClient()
+        resp = client.post(url, headers=headers, params=params, data=data, json=json)
```
[the remark this answers is not in the document]
Comment (kgang) [reply concrete/tools.py:145]: I think that's up to the person calling this to know. There's a `raise_for_status` if the request format is bad

File: concrete/tools.py (hunk 1 above)
Comment (MichaelDeng03) [reply concrete/tools.py]: Probably could add source and target branch args instead of hard coding main

**Review by MichaelDeng03 — COMMENTED** [review]
left a couple of comments

**Review by MichaelDeng03 — APPROVED** [review]

**Pull request merged**